✨ show edited text messages

Chun-Cheng opened the pull request:
- Show edited messages, including the original sending time and an "edited" label.
- Preserve for other message types.

Ref MES-113

Chun-Cheng commented on src/api/message/message-service.ts:
```diff
@@ -57,47 +57,46 @@ export const SendMessage = async (roomId: bigint, author: bigint, text: string):
  * @returns An async generator that yields messages. Throws an error if the request fails.
  */
-export const GetMessage = async function* (roomId: bigint, baseline?: bigint, limit?: number): AsyncGenerator<Message, void, void> {
+export const GetLatestMessage = async function* (roomId: bigint, userId : bigint|undefined): AsyncGenerator<Message, void , void> {
```
The added space after `void` is not necessary. (Apparently our ESLint settings is not powerful enough to catch this.)

Chun-Cheng committed:
🚧 support text format

Chun-Cheng committed:
✨ shows edited text message

Shows original sending time and last editing time.

Chun-Cheng committed:
♻️ rename message-text.tsx to text-message.tsx

Chun-Cheng committed:
🐛 fix copilot suggested changes

- update the comment of sorting messages
- remove console.log all messages
- add slice to the return value in the catch block

Chun-Cheng committed:
♻️ use format to decide whether to use formatted body or not

follow the Matrix spec

hungsean reviewed (commented):
老實說
程式碼有點超出我認知範圍
先寫一下目前觀察到的問題
1. 其他client編輯過後，雖然會即時顯示，但會是以新訊息的狀態顯示，時間會寫invalid time
![image](https://github.com/user-attachments/assets/1344fc81-6cde-49a2-be66-1caeae4fa4c6)
2. 編輯後訊息會有個 \* 我不知道是故意的還是有什麼問題

然後我請codex review 了一下

---

以下是針對 `cheng/message-types` 分支的 PR Review：

### 優點

* 將富文字的渲染邏輯抽離至 `TextMessage` 元件，使 `MessageItem` 更專注於整體消息結構，讓架構更清晰易維護。
* 在客戶端分頁時能正確合併「編輯」事件，並以最新內容取代原始消息，達成預期行為。
* UI 上新增了 `edited` 標記，明確告知使用者該則訊息已被修改，提升可用性。

---

### 建議優化的部分

#### 1. `TextMessage` 元件的安全性問題

* 目前直接使用 `dangerouslySetInnerHTML`，未進行任何 HTML 淨化處理。建議引入 [[DOMPurify](https://github.com/cure53/DOMPurify)](https://github.com/cure53/DOMPurify) 或其他類似庫，以白名單方式過濾內容，避免潛在的 XSS 攻擊。
* 建議先檢查 `content.formatted_body` 是否存在且非空，若無則退回至 `content.body`，以防第三方僅傳送 `format` 欄位卻漏掉 `formatted_body`。

#### 2. `getRoomMessages` 中「編輯」事件的處理重複且易出錯

* 初次處理編輯事件時會給消息加上 `_originalTs`，但在分頁載入時未保留這個邏輯，導致後續無法正確呈現原始時間。
* 建議抽出一個 helper 函式，如 `mergeEdits(events: MatrixEvent[]): MatrixEvent[]`，統一處理流程，包括：

  1. 依時間順序（或用 `getTs()` 排序）
  2. 篩選 `m.replace` 類型的關聯事件
  3. 補上 `_originalTs`
  4. 使用 Map 等結構保留每則消息的最新版本
* 初始載入與分頁後都共用這段邏輯，可降低重複與錯誤風險。

#### 3. 排序與截斷邏輯易導致選到最舊消息

* 目前採用 `events.sort((a, b) => b.ts - a.ts)`（由新至舊排序）後再用 `slice(-limit)`，實際上會拿到**最舊的 N 筆資料**，與「取最新 N 筆」的預期相反。
* 建議修改為：

  * 先升序排序：`events.sort((a, b) => a.ts - b.ts)` → 由舊到新
  * 再取後段：`events.slice(-limit)` → 最新 N 筆（順序為舊 → 新）
* 或反過來：先倒序後 `slice(0, limit)` 取最新，再根據 UI 需要 reverse 回原本順序。

#### 4. `MessageItem` 對 `msgtype` 的支援尚不完整

* 除了 `m.text`，目前對其他訊息類型（如圖片、文件、emote、notice 等）一律回傳空節點 `<></>`，導致這些訊息無法顯示。
* 建議至少先支援基本 fallback，例如：

  * `m.notice`、`m.emote` 可顯示簡單樣式的文字：
    `<div className="italic text-sm">{content.body}</div>`
  * 圖片、影片等類型可先顯示占位訊息（如：「\[圖片訊息]」），後續再開發專用元件。
* 建議將目前多層巢狀的三元運算式改為 `switch` 或 `msgtype → 元件` 的對映表，提升可讀性與擴展性。

#### 5. 其他細節建議

* `sendTextMessage` 中使用 `client.client.sendTextMessage`，命名略顯冗長。可考慮將導出的 `client` 改為 `matrixClient`，或調整內部命名以降低混淆。
* `MessageItem` 雖有引入 `getUserAvatar`，但使用場景有限，確保僅在 `AvatarImage` 中謹慎調用即可。
* 建議為 `getRoomMessages` 補上單元測試，覆蓋以下情境：

  * 僅有普通訊息的情況
  * 單一與多個編輯事件的處理
  * 分頁後仍能正確合併編輯版本

---

### 總結

整體來說，這個 PR 已成功實作富文字渲染與編輯合併的核心需求。但仍有排序錯誤與分頁後時間資訊遺失等問題，可能會在實際操作中造成用戶困擾。若能依據以上建議進一步細化重構，即可考慮合併此分支。

Chun-Cheng committed:
♻️ refactor edited mesages logic

Create the TimelineItem class to store the message state changes.
Reference: matrix-rust-sdk matrix-sdk-ui crate
- https://github.com/matrix-org/matrix-rust-sdk/blob/main/crates/matrix-sdk-ui/src/timeline/item.rs
- https://github.com/matrix-org/matrix-rust-sdk/blob/main/crates/matrix-sdk-ui/src/timeline/event_handler.rs
